spec/requests/api/v3/esdl_files_controller_spec.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe 'ESDL files', :etsource_fixture do
+  # rubocop:disable RSpec/BeforeAfterAll
+  before(:all) do
+    NastyCache.instance.expire!
+  end
+  # rubocop:enable RSpec/BeforeAfterAll
+
+  let(:scenario) { FactoryBot.create(:scenario) }
+  let(:url) { "/api/v3/scenarios/#{scenario.id}/esdl_file" }
+
+  context('when requesting attached esdl file') do
+    context 'when nothing is attached' do
+      before { get(url) }
+
+      it 'succeeds' do
+        expect(response).to be_successful
+      end
+
+      it 'sends no esdl data' do
+        expect(JSON.parse(response.body)).to eq({})
+      end
+    end
+
+    context 'when file is attached' do
+      before do
+        put url, params: {
+          file: fixture_file_upload('files/valid_esdl_file.esdl', 'text/xml').read,
+          filename: 'my esdl'
+        }
+
+        get(url)
+      end
+
+      it 'succeeds' do
+        expect(response).to be_successful
+      end
+
+      it 'sends 2 pieces of data about the esdl file' do
+        expect(JSON.parse(response.body).length).to eq(2)
+      end
+
+      it 'sends data about the esdl file' do
+        expect(JSON.parse(response.body)).to include(
+          'filename' => 'my esdl'
+        )
+      end
+    end
+
+    context 'with download parameter set' do
+      before do
+        put url, params: {
+          file: fixture_file_upload('files/valid_esdl_file.esdl', 'text/xml').read,
+          filename: 'my esdl'
+        }
+
+        get(url, params: { download: true })
+      end
+
+      it 'succeeds' do
+        expect(response).to be_successful
+      end
+
+      it 'sends 3 pieces of data about the esdl file' do
+        expect(JSON.parse(response.body).length).to eq(3)
+      end
+
+      it 'sends data about the esdl file' do
+        expect(JSON.parse(response.body)).to include(
+          'filename' => 'my esdl'
+        )
+      end
+
+      it 'sends the full esdl file' do
+        expect(JSON.parse(response.body)['file']).to eq(
+          fixture_file_upload('files/valid_esdl_file.esdl', 'text/xml').read
+        )
+      end
+    end
+  end
+
+  context('when uploading a valid esdl file') do
+    let(:request) do
+      put url, params: {
+        file: fixture_file_upload('files/valid_esdl_file.esdl', 'text/xml').read,
+        filename: 'my esdl'
+      }
+    end
+
+    it 'succeeds' do
+      request
+      expect(response).to be_successful
+    end
+
+    it 'attaches the file' do
+      expect { request }
+        .to change {
+          scenario
+            .reload
+            .attachments
+            .find_by(key: 'esdl_file')
+            .present?
+        }
+        .from(false).to(true)
+    end
+  end
+
+  context('when uploading an esdl file larger than 5MB') do
+    let(:request) do
+      put url, params: {
+        file: fixture_file_upload('files/huge_invalid_esdl_file.esdl', 'text/xml').read,
```

Comment:
Could we [generate this file dynamically](https://github.com/quintel/etengine/blob/05e87d469e2d2e2011231e17ceadc6508a30757d/spec/requests/api/v3/custom_curve_spec.rb#L230-L240) in the test, rather than storing a 5MB file in the repo, or would that defeat the point of the test? I guess this is harder to do for ESDL than it is for a simple curve file.

Reply:
It shouldn't be so hard if we don't check for validity. We only need the xml and ESDL headers and closing brackets. Then we can just fill it up with an endless amount of wind turbine assets.

Reply:
There's now a dynamically generated file containing 25,000 wind turbines!